src/java_bytecode/generate_java_generic_type.cpp

```diff
@@ -0,0 +1,102 @@
+/*******************************************************************\
+
+ Module: Generate Java Generic Type - Instantiate a generic class with
+         concrete type information.
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+#include "generate_java_generic_type.h"
+#include <util/namespace.h>
+#include <java_bytecode/java_types.h>
+#include <java_bytecode/java_utils.h>
+
+#include <iostream>
+#include <sstream>
+#include <algorithm>
+
+generate_java_generic_typet::generate_java_generic_typet(
+  message_handlert &message_handler):
+    message_handler(message_handler)
+{}
+
+/// Generate a concrete instantiation of a generic type.
+/// \param existing_generic_type The type to be concretised
+/// \param symbol_table The symbol table that the concrete type will be inserted into.
+/// \return The symbol as it was retrieved from the symbol table after it has been inserted into.
+symbolt generate_java_generic_typet::operator()(
+  const java_generic_typet &existing_generic_type,
+  symbol_tablet &symbol_table) const
+{
+  PRECONDITION(existing_generic_type.id()==ID_pointer);
+  namespacet ns(symbol_table);
+
+  const typet &pointer_subtype=ns.follow(existing_generic_type.subtype());
+
+  INVARIANT(
+    pointer_subtype.id()==ID_struct, "Only pointers to classes in java");
+
+  java_class_typet &replacement_type=to_java_class_type(const_cast<typet &>(pointer_subtype));
+  const irep_idt new_tag=build_generic_tag(existing_generic_type, replacement_type);
+  replacement_type.set_tag(new_tag);
+  
+  // Small auxiliary function, to perform the inplace modification of the generic fields.
+  auto replace_type_for_generic_field=
+    [&](struct_union_typet::componentt &component)
+    {
+      if(is_java_generic_parameter(component.type()))
+      {
+        INVARIANT(
```

Comment:
@thk123 This becomes way more readable if I remove the invariant from here. I wanted to ask if that makes sense (it's guarding against a case that's always guaranteed to be true in this case).

Reply:
Invariants are _always_ meant to be guarding against cases that are guaranteed to be true? 

I'm also not sure that is true in this case, if it is, why is `generic_type_variables` a vector data structure vs. just a single element? I think I put this in to remind me to deal with when we have generic classes with multiple generic parameters.